File: nanoGPT/kan/bspline.py
"""B-spline grid and basis function evaluation using PyTorch.

Provides uniform cubic B-spline grids with Cox-de Boor evaluation,
Greville abscissae for identity initialization, and dynamic grid expansion.
All operations are vectorized over batch dimensions using torch tensors.
"""

import math
import torch
import torch.nn.functional as F


class BSplineGrid:
    """Uniform B-spline grid with vectorized basis evaluation.

    The grid covers [grid_min, grid_max] with uniform knot spacing.
    Supports dynamic expansion when inputs fall outside the grid range.
    """

    def __init__(self, order: int, num_basis: int, grid_min: float, grid_max: float):
        self.order = order
        self.num_basis = num_basis
        self.grid_min = grid_min
        self.grid_max = grid_max

        num_interior = max(num_basis - order, 1)
        self.step = (grid_max - grid_min) / num_interior

        num_knots = num_basis + order + 1
        self.knots = torch.tensor(
            [grid_min + (i - order) * self.step for i in range(num_knots)],
            dtype=torch.float32,
        )

    def evaluate(self, x: torch.Tensor) -> torch.Tensor:
        """Evaluate all basis functions at points x.

        Args:
            x: tensor of shape (...,) — any batch shape

        Returns:
            Tensor of shape (..., num_basis) — basis function values
        """
        flat = x.reshape(-1)
        t = self.knots
        k = self.order + 1  # degree + 1
        num_intervals = len(t) - 1

        # Degree-0 basis: piecewise constant
        # basis[i] = 1 if t[i] <= x < t[i+1], with right endpoint included for last interval
        t_lo = t[:num_intervals]  # (num_intervals,)
        t_hi = t[1 : num_intervals + 1]  # (num_intervals,)

        x_exp = flat.unsqueeze(-1)  # (N, 1)
        basis = ((x_exp >= t_lo) & (x_exp < t_hi)).float()  # (N, num_intervals)

        # Include right endpoint for the last interval
        basis[:, -1] += (flat == t[-1]).float()

        # Cox-de Boor recursion for degrees 1..order
        for d in range(1, k):
            n_new = num_intervals - d
            t_left = t[:n_new]
            t_right = t[d : d + n_new]
            denom1 = t_right - t_left
            safe1 = denom1.clamp(min=1e-10)
            left = (x_exp - t_left) / safe1 * basis[:, :n_new]
            left = left * (denom1 > 0).float()

            t_left2 = t[1 : n_new + 1]
            t_right2 = t[d + 1 : d + 1 + n_new]
            denom2 = t_right2 - t_left2
            safe2 = denom2.clamp(min=1e-10)
            right = (t_right2 - x_exp) / safe2 * basis[:, 1 : n_new + 1]
            right = right * (denom2 > 0).float()

            basis = left + right

        # Trim or pad to num_basis
        if basis.shape[-1] > self.num_basis:
            basis = basis[:, : self.num_basis]
        elif basis.shape[-1] < self.num_basis:
            basis = F.pad(basis, (0, self.num_basis - basis.shape[-1]))

        return basis.reshape(*x.shape, self.num_basis)

    def greville_abscissae(self) -> torch.Tensor:
        """Compute Greville abscissae — coefficients that make KAN(x) = x.

        The Greville abscissa for basis function i is the average of k-1
        consecutive knots starting at position i+1. When used as B-spline
        coefficients, the resulting curve is the identity function.
        """
        k = self.order + 1
        coeffs = torch.zeros(self.num_basis)
        for i in range(self.num_basis):
            indices = [i + j for j in range(1, k) if i + j < len(self.knots)]
            if indices:
                coeffs[i] = self.knots[indices].mean()
        return coeffs

    def expand(self, new_min: float, new_max: float) -> tuple["BSplineGrid", int]:
        """Expand grid to cover [new_min, new_max], preserving step size.

        Returns (new_grid, left_offset) where left_offset is the number
        of new basis functions added on the left. Old basis i maps to
        new basis (i + left_offset).
        """
        left_steps = 0
        if new_min < self.grid_min:
            left_steps = math.ceil((self.grid_min - new_min) / self.step)
        right_steps = 0
        if new_max > self.grid_max:
            right_steps = math.ceil((new_max - self.grid_max) / self.step)

        if left_steps == 0 and right_steps == 0:
            return self, 0

        expanded_min = self.grid_min - left_steps * self.step
        expanded_max = self.grid_max + right_steps * self.step
        new_num_basis = self.num_basis + left_steps + right_steps

        new_grid = BSplineGrid(self.order, new_num_basis, expanded_min, expanded_max)
        new_grid.step = self.step  # Force same step to avoid floating-point drift
        return new_grid, left_steps

File: ml/nn/kan/bspline.go
package kan

import "math"

// BSplineGrid holds a uniform knot vector and evaluates cubic B-spline basis functions.
// All computation is done in pure Go on CPU since the coefficient tensors are tiny.
//
// The grid can be dynamically expanded via Expand() when logits fall outside the
// current [GridMin, GridMax] range. Expansion preserves the step size and existing
// knot positions, so previously-evaluated basis functions are unchanged.
type BSplineGrid struct {
	Order    int
	NumBasis int
	Knots    []float32 // length = NumBasis + Order + 1
	GridMin  float32   // lower bound of the active region
	GridMax  float32   // upper bound of the active region
	Step     float32   // uniform spacing between interior knots
}

// NewBSplineGrid creates a uniform B-spline grid with the given parameters.
// For cubic B-splines (order=3) with numBasis=8, this creates 12 knots
// uniformly spaced in [gridMin, gridMax] with extended boundary knots.
func NewBSplineGrid(order, numBasis int, gridMin, gridMax float32) *BSplineGrid {
	numInterior := numBasis - order
	if numInterior < 1 {
		numInterior = 1
	}

	step := (gridMax - gridMin) / float32(numInterior)
	numKnots := numBasis + order + 1
	knots := make([]float32, numKnots)

	for i := range knots {
		knots[i] = gridMin + float32(i-order)*step
	}

	return &BSplineGrid{
		Order:    order,
		NumBasis: numBasis,
		Knots:    knots,
		GridMin:  gridMin,
		GridMax:  gridMax,
		Step:     step,
	}
}

// Expand returns a new BSplineGrid covering [newMin, newMax] with the same step
// size and order. Boundaries are snapped outward to step multiples so the knot
// vector stays uniform.
//
// Returns the new grid and a leftOffset: old basis function i maps to new basis
// function (i + leftOffset). This offset is needed to reposition existing
// coefficients in the expanded weight vector.
//
// If the current grid already covers [newMin, newMax], returns (self, 0).
func (g *BSplineGrid) Expand(newMin, newMax float32) (*BSplineGrid, int) {
	// How many steps to extend in each direction?
	leftSteps := 0
	if newMin < g.GridMin {
		leftSteps = int(math.Ceil(float64(g.GridMin-newMin) / float64(g.Step)))
	}
	rightSteps := 0
	if newMax > g.GridMax {
		rightSteps = int(math.Ceil(float64(newMax-g.GridMax) / float64(g.Step)))
	}

	if leftSteps == 0 && rightSteps == 0 {
		return g, 0
	}

	expandedMin := g.GridMin - float32(leftSteps)*g.Step
	expandedMax := g.GridMax + float32(rightSteps)*g.Step
	newNumBasis := g.NumBasis + leftSteps + rightSteps

	newGrid := NewBSplineGrid(g.Order, newNumBasis, expandedMin, expandedMax)
	// Force the same step to avoid floating-point drift
	newGrid.Step = g.Step

	return newGrid, leftSteps
}

// Evaluate computes the values of all basis functions at a single point x.
// Uses the Cox-de Boor recursion algorithm.
// Returns a slice of length NumBasis.
func (g *BSplineGrid) Evaluate(x float32) []float32 {
	k := g.Order + 1 // B-spline order (degree + 1), e.g., 4 for cubic
	n := g.NumBasis
	t := g.Knots

	// Start with degree-0 basis functions (piecewise constant)
	numIntervals := len(t) - 1
	basis := make([]float32, numIntervals)
	for i := 0; i < numIntervals; i++ {
		if (x >= t[i] && x < t[i+1]) || (i == numIntervals-1 && x == t[i+1]) {
			basis[i] = 1.0
		}
	}

	// Cox-de Boor recursion for degrees 1..order
	for d := 1; d < k; d++ {
		newBasis := make([]float32, numIntervals-d)
		for i := range newBasis {
			var left, right float32

			denom1 := t[i+d] - t[i]
			if denom1 > 0 {
				left = (x - t[i]) / denom1 * basis[i]
			}

			denom2 := t[i+d+1] - t[i+1]
			if denom2 > 0 {
				right = (t[i+d+1] - x) / denom2 * basis[i+1]
			}

			newBasis[i] = left + right
		}
		basis = newBasis
	}

	// Ensure we return exactly NumBasis values
	if len(basis) > n {
		basis = basis[:n]
	}
	for len(basis) < n {
		basis = append(basis, 0)
	}

	return basis
}

// EvaluateBatch computes basis function values for a batch of input points.
// Returns a [len(xs)][NumBasis] matrix (row-major).
func (g *BSplineGrid) EvaluateBatch(xs []float32) [][]float32 {
	result := make([][]float32, len(xs))
	for i, x := range xs {
		result[i] = g.Evaluate(x)
	}
	return result
}

// InitSoftmaxApprox returns initial B-spline coefficients that make the KAN
// approximate the identity function: kan(x) ≈ x.
//
// Since the forward pass applies exp(kan(x) - rowMax) / sum(exp(...)),
// initializing to identity means the full pipeline computes:
//
//	exp(x - max) / sum(exp(...)) ≈ softmax(x)
//
// This gives the KAN a near-perfect starting point, and training refines
// it from there.
//
// The coefficients are the Greville abscissae of the knot vector, which is
// the standard way to make a B-spline curve interpolate the identity. No
// further normalization is applied — shifting or rescaling the coefficients
// would change the effective slope away from 1.0 and break the identity
// approximation.
func InitSoftmaxApprox(grid *BSplineGrid) []float32 {
	n := grid.NumBasis
	coeffs := make([]float32, n)

	// To approximate identity with B-splines, the coefficients should be
	// the x-coordinates at each basis function's center (Greville abscissae).
	// For a uniform knot vector, these are evenly spaced.
	k := grid.Order + 1 // degree + 1
	for i := 0; i < n; i++ {
		// Greville abscissa: average of k-1 consecutive knots starting at i+1
		sum := float32(0)
		count := 0
		for j := 1; j < k; j++ {
			idx := i + j
			if idx < len(grid.Knots) {
				sum += grid.Knots[idx]
				count++
			}
		}
		if count > 0 {
			coeffs[i] = sum / float32(count)
		}
	}

	return coeffs
}
